controllers/twophase/types.go

```diff
@@ -61,206 +61,80 @@ func (r *Reconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
 	}
 	chaos := _chaos.(v1alpha1.InnerSchedulerObject)
 
-	duration, err := chaos.GetDuration()
-	if err != nil {
-		r.Log.Error(err, "failed to get chaos duration")
-		return ctrl.Result{}, err
+	status := chaos.GetStatus()
+
+	targetPhase := status.Experiment.Phase
+
+	if !chaos.GetNextRecover().IsZero() && chaos.GetNextRecover().Before(now) {
+		targetPhase = v1alpha1.ExperimentPhaseWaiting
 	}
 
-	scheduler := chaos.GetScheduler()
-	if scheduler == nil {
-		r.Log.Info("Scheduler should be defined currently")
-		return ctrl.Result{}, fmt.Errorf("misdefined scheduler")
+	if chaos.GetNextStart().Before(now) {
+		targetPhase = v1alpha1.ExperimentPhaseRunning
 	}
 
-	if duration == nil {
-		zero := 0 * time.Second
-		duration = &zero
+	if chaos.IsPaused() {
+		targetPhase = v1alpha1.ExperimentPhasePaused
 	}
 
-	status := chaos.GetStatus()
+	// TODO: find a better way to solve the pause and resume problem.
+	// Or pause is a bad design for the scheduler :(
+	if !chaos.IsPaused() && status.Experiment.Phase == v1alpha1.ExperimentPhasePaused {
```

Comment:
Little confusion here.

At previous commit, each if-else, e.g. `!chaos.GetNextRecover().IsZero() && chaos.GetNextRecover().Before(now)` and `!chaos.IsPaused() && status.Experiment.Phase == v1alpha1.ExperimentPhasePaused` wouldn't run at the same time, but now, it did.

Think about the following case:

1. a chaos created and started;
2. it has been paused before nextRecover reached;
3. nextRecover reached but nothing changed since the chaos is still been `Paused` :(
4. the chaos has been unpaused

and now, both `!chaos.GetNextRecover().IsZero() && chaos.GetNextRecover().Before(now)` and `!chaos.IsPaused() && status.Experiment.Phase == v1alpha1.ExperimentPhasePaused` would be true, the targetPhase become `ExperimentPhaseRunning`!

Although it would not make the chaos really run `apply`, because [Line 149] if check would be false, and the targetPhase would be changed to `ExperimentPhaseWaiting` after did `resume`.

Is this acceptable?

Reply:
This implement keeps unifying with the current version, you can check here https://github.com/chaos-mesh/chaos-mesh/blob/master/controllers/twophase/types.go#L138. I think this implement is acceptable at present. If we want to improve this, we can do it in another pr.

Reply:
I think it's not the same case, what you mean in L138 should be `started -> paused -> now -> nextRecover -> nextStart`, and when I unpaused at **now**, it do retrying.

But what I mean is that `started -> paused -> nextRecover -> now -> nextStart`, and when I unpaused at **now**, it will enter both [Line 68](https://github.com/chaos-mesh/chaos-mesh/pull/1221/files#diff-04c792dfed57630c1de983400da2e7639c6d5c3a929c0124f7b5c87aa9a19105R68) and [Line 82](https://github.com/chaos-mesh/chaos-mesh/pull/1221/files#diff-04c792dfed57630c1de983400da2e7639c6d5c3a929c0124f7b5c87aa9a19105R72), and targetPhase become `ExperimentPhaseRunning` at this moment 😅

And then,  it enter `resume()` [Line 156](https://github.com/chaos-mesh/chaos-mesh/pull/1221/files#diff-d396848e41155ba5b591cf171a8ab550845161333a956422f14c840b611d16bcR156).

It's not a big problem since it's finally do `noop`, but it still will leave one line log say:

```
2020-12-11T10:25:24.033Z	INFO	controllers.networkchaos.default/web-show-network-delay	change phase	{"reconciler": "networkchaos", "resource name": "default/web-show-network-delay", "current phase": "Paused", "target phase": "Running"}
```

and then turn to `Waiting` soon:

```
2020-12-11T10:25:24.051Z	INFO	controllers.networkchaos.default/web-show-network-delay	change phase	{"reconciler": "networkchaos", "resource name": "default/web-show-network-delay", "current phase": "Waiting", "target phase": "Waiting"}
```

Reply:
Yes. It's the dirty part. I have commented that target `Running` and `Recovering` has the same behavior for `resume`. I haven't come up with a good idea on how to solve the problem 😭

Reply:
I always think this is a feature that you could see the chaos experiments being resumed IMMEDIATELY after clicking unpause.

Reply:
If we did think so, how about make this targetPhase to a temporary phase, e.g. v1alpha1.ExperimentPhaseResuming   😁

Reply:
@STRRL This phase will not be updated to `Kubernetes`. As you can see in the state machine, though we passed `targetPhase` as `Running`, it will be updated to `Waiting` (or `Failed`). It can only be seen in the log.